linalg/constraints.cpp

```diff
@@ -0,0 +1,893 @@
+}
+
+SchurConstrainedHypreSolver::~SchurConstrainedHypreSolver()
+{
+   delete schur_mat;
+   delete primal_pc;
+}
+#endif
+
+void ConstrainedSolver::Initialize()
+{
+   height = A.Height() + B.Height();
+   width = A.Width() + B.Height();
+
+   workb.SetSize(A.Height());
+   workx.SetSize(A.Height());
+   constraint_rhs.SetSize(B.Height());
+   constraint_rhs = 0.0;
+   multiplier_sol.SetSize(B.Height());
+}
+
+#ifdef MFEM_USE_MPI
+ConstrainedSolver::ConstrainedSolver(MPI_Comm comm, Operator& A_, Operator& B_)
+   :
+   IterativeSolver(comm), A(A_), B(B_)
+{
+   Initialize();
+}
+#endif
+
+ConstrainedSolver::ConstrainedSolver(Operator& A_, Operator& B_)
+   :
+   A(A_), B(B_)
+{
+   Initialize();
+}
+
+void ConstrainedSolver::SetConstraintRHS(const Vector& r)
+{
+   MFEM_VERIFY(r.Size() == multiplier_sol.Size(), "Vector is wrong size!");
+   constraint_rhs = r;
+}
+
+void ConstrainedSolver::PrimalMult(const Vector& f, Vector &x) const
+{
+   Vector pworkb(A.Height() + B.Height());
+   Vector pworkx(A.Height() + B.Height());
+   pworkb = 0.0;
+   pworkx = 0.0;
+   for (int i = 0; i < f.Size(); ++i)
+   {
+      pworkb(i) = f(i);
+      pworkx(i) = x(i);
+   }
+   for (int i = 0; i < B.Height(); ++i)
+   {
+      pworkb(f.Size() + i) = constraint_rhs(i);
+   }
+
+   Mult(pworkb, pworkx);
+
+   for (int i = 0; i < f.Size(); ++i)
+   {
+      x(i) = pworkx(i);
+   }
+   for (int i = 0; i < B.Height(); ++i)
+   {
+      multiplier_sol(i) = pworkx(f.Size() + i);
+   }
+}
+
+void ConstrainedSolver::Mult(const Vector& f_and_r, Vector& x_and_lambda) const
+{
+   workb.MakeRef(const_cast<Vector&>(f_and_r), 0);
+   workx.MakeRef(x_and_lambda, 0);
+   Vector ref_constraint_rhs(f_and_r.GetData() + A.Height(), B.Height());
+   constraint_rhs = ref_constraint_rhs;
+   PrimalMult(workb, workx);
+   Vector ref_constraint_sol(x_and_lambda.GetData() + A.Height(), B.Height());
+   GetMultiplierSolution(ref_constraint_sol);
+}
+
+/* Helper routine to reduce code duplication - given a node (which MFEM
+   sometimes calls a "dof"), this returns what normal people call a dof but
+   which MFEM sometimes calls a "vdof" - note that MFEM's naming conventions
+   regarding this are not entirely consistent. In parallel, this always
+   returns the "truedof" in parallel numbering. */
+int CanonicalNodeNumber(FiniteElementSpace& fespace,
+                        int node, bool parallel, int d=0)
+{
+#ifdef MFEM_USE_MPI
+   if (parallel)
+   {
+      try
+      {
+         ParFiniteElementSpace& pfespace =
+            dynamic_cast<ParFiniteElementSpace&>(fespace);
+         const int vdof = pfespace.DofToVDof(node, d);
+         return pfespace.GetLocalTDofNumber(vdof);
+      }
+      catch (std::bad_cast&)
+      {
+         MFEM_ABORT("Asked for parallel form of serial object!");
+         return -1;
+      }
+   }
+   else
+#endif
+   {
+      return fespace.DofToVDof(node, d);
+   }
+}
+
+SparseMatrix * BuildNormalConstraints(FiniteElementSpace& fespace,
```

Comment:
Thanks for updating the comments within here. It's much easier to follow what's going within the code now :)